tests/functional/test_install_direct_url.py

```diff
@@ -46,3 +46,14 @@ def test_install_archive_direct_url(script, data, with_wheel):
     assert req.startswith("simple @ file://")
     result = script.pip("install", req)
     assert _get_created_direct_url(result, "simple")
+
+
+def test_install_vcs_constraint_direct_url(script, with_wheel):
+    constraints_file = script.scratch_path / "constraints.txt"
+    constraints_file.write_text(
+        "git+https://github.com/pypa/pip-test-package"
+        "@5547fa909e83df8bd743d3978d6667497983a4b7"
+        "#egg=pip-test-package"
+    )
+    result = script.pip("install", "pip-test-package", "-c", constraints_file)
+    assert _get_created_direct_url(result, "pip_test_package")
```

Comment:
This needs a `@pytest.mark.network` marker. I’d also add a test for `git+file://` as well that relies on a `data.packages` entry.